windows/src/engine/kmcomapi/com/system/keymancontrol.pas

```diff
@@ -359,47 +358,20 @@ function TKeymanControl.LastRefreshToken: IntPtr;
   Result := FLastRefreshToken;
 end;
 
+type
+  TRefreshThread = class(TThread)
+  private
+    FLastRefreshToken: Integer;
+  protected
+    procedure Execute; override;
+  public
+    constructor Create(ALastRefreshToken: Integer);
+  end;
+
 procedure TKeymanControl.ApplyToRunningKeymanEngine;
 begin
-  // This convoluted way of refreshing keyman ensures that km is init for the thread.
-  // Other methods would work but this is easiest
-  TThread.CreateAnonymousThread(
-    procedure
-    const
-      KR_REQUEST_REFRESH = 0;
-      KR_SETTINGS_CHANGED = 3;
-    var
-      msg: TMsg;
-      wm_keyman_refresh: UINT;
-    begin
-      wm_keyman_refresh := RegisterWindowMessage('WM_KEYMANREFRESH');
-      RefreshHandle := AllocateHWnd(nil);
-
-      // We currently have two announcements because the KR_REQUEST_REFRESH
-      // announcement is actioned only when Keyman Engine is running, while we
-      // still need to tell any apps listening for changes that the settings
-      // have changed. In future, we should probably refactor this to use a
-      // single broadcast.
-
-      // We use a random number here to so that multiple processes can generate
-      // hopefully unique tokens for refresh, and we share this token with our
-      // current consumer so they can ignore notifications that they have
-      // generated
-
-      // First, post out a KR_SETTINGS_CHANGED for the benefit of Keyman
-      // Configuration and other apps that want it
-      FLastRefreshToken := Random(MaxInt);
-      PostMessage(HWND_BROADCAST, wm_keyman_refresh, KR_SETTINGS_CHANGED, FLastRefreshToken);
-
-      // Then, post out a refresh to Keyman Engine and process it in this
-      // thread so that Keyman Engine will grab it and broadcast it, if it is
-      // currently running
-      PostMessage(RefreshHandle, wm_keyman_refresh, KR_REQUEST_REFRESH, 0);
-      GetMessage(msg, RefreshHandle, wm_keyman_refresh, wm_keyman_refresh);
-      DispatchMessage(msg);
-      DeallocateHWnd(RefreshHandle);
-    end
-  ).Start;
+  FLastRefreshToken := Random(MaxInt);
```

Comment:
... a _random_ token?  Why not seed it from an incrementing value?  Is that because it needs to stay unique, even across multiple threads?

If so, what about a rotating token scheme?  Threads usually have some sort of ID value (or at least, can be assigned one), so defining a legal 'window' for each threads token value and using modular arithmetic should be able to address the issue as well.  Just use the thread's ID to define the range's root and set a constant width for the window.

With the random token, I'd worry that a random number generator might produce equal values very close in time to each other.  Sure, it's unlikely... but possible.  Those rare cases would then be _pretty_ tricky to debug if they caused errors.

Reply:
> Why not seed it from an incrementing value? Is that because it needs to stay unique, even across multiple threads?

Yes, it needs to be unique across threads and it needs to be generated independently. We have a single 32 bit value we can use here. 

> If so, what about a rotating token scheme? Threads usually have some sort of ID value (or at least, can be assigned one), so defining a legal 'window' for each threads token value and using modular arithmetic should be able to address the issue as well. Just use the thread's ID to define the range's root and set a constant width for the window.

A thread ID is an arbitrary 32 bit value and the token is also a 32 bit value. So I think we'd still be in the same boat, wouldn't we?

> Sure, it's unlikely... but possible. Those rare cases would then be _pretty_ tricky to debug if they caused errors.

I figured it was vanishingly rare and with subtle enough side-effects that we'll probably never see it happen. This seemed to me to be the simplest way to have reasonable guarantees of token uniqueness across threads without needing synchronisation